### ISSUE TITLE
Questions about Meta-RL evaluation metrics

### ISSUE THREAD

**Issue by yjy1231** [rainx0r/metaworld-algorithms#10]
I would like to raise a question about the evaluation metric of Meta-RL environments. Taking ML1-reachv3 as an example (ML10/45 have similar problems). From the implementation, the success rate is counted as `n_successes / n_evals`, where `n_evals` has a default value of 3. The question I'm having is how `n_successes` is calculated. Intuitively, once a success is achieved in a testing environment, that testing environment should not be contributing to the success rate any more. However, from the code and the results, I can tell that the same testing environment contributes to the success rate multiple times. For example, consider testing batch size 10, and 9 out of 10 envs fail, but one was very successful and finished 3 runs within the max_step. This successful environment will contribute 3 successes, hence getting a 100% success rate while the actual success rate was only 10%. This way of calculating the success rate has significantly overestimated the success rate. 

I tried to fix this by setting `n_evals` equal to the testing batch size, and have flags to indicate whether an environment has contributed to the success rate already. Testing this on RL2 gives only <50% success rate at 1M training steps.

**Comment by rainx0r** [comment]
The evaluation scheme currently implemented is that from the original metaworld publication code in the [garage repo](https://github.com/rlworkgroup/garage/blob/master/src/garage/experiment/meta_evaluator.py#L10). It's also documented in the [Metaworld docs](https://metaworld.farama.org/evaluation/evaluation/#meta-reinforcement-learning-ml1-ml10-ml45). It's also worth nothing that we don't implement any evaluation methodology in this repo, it's all done [in the metaworld package itself](https://github.com/Farama-Foundation/Metaworld/blob/master/metaworld/evaluation.py#L106) so evaluation-related issues should probably go there.

To address your actual question, the success rate for meta RL isn't really `n_successes / n_evals`, it's actually `success_rate / n_evals`. Essentially how it works is kind of like this (again there's far more documentation in the links provided):

```py
total_success_rate = 0
for i in range(num_evals):
    envs.sample_goals()
    adapted_agent = agent.copy()
    for j in range(adaptation_steps):
        data = rollout(adaptation_episodes, adapted_agent, envs)
        adapted_agent = adapted_agent.adapt(data)
    # Multitask evaluation will evaluate the adapted agent for num_eval_episodes on each env (with a frozen goal position)
    # and return the success rate first averaged within each env, then averaged across envs
    success_rate_after_adaptation = multitask_evaluation(adapted_agent, envs, num_eval_episodes)
    total_success_rate += success_rate_after_adaptation
metarl_success_rate = total_success_rate / num_evals
```
Note the comment. Let's say that `envs` has 3 envs and `num_evals` is 2 and `num_eval_episodes` is 5.
During the first evaluation (`i=0`), let's say the agent does really well on the first env only and gets it right all 5 post-adaptation episodes, then for env 2 it didn't adapt as well and only got 3/5 episodes, and env 3 It didn't get it at all and got 0/5. In this case, `success_rate_after_adaptation == (1.0 + 3/5 + 0) / 3` which is `0.5`. Then let's say in the second evaluation batch too  (`i=1`) after adaptation the agent gets 4/5 in the first env, 2/5 in the second env, 1/5 in the third env, then `success_rate_after_adaptation == (4/5 + 2/5 + 1/5) / 3` or `0.46`. Then `metarl_success_rate == (0.5 + 0.46) / 2` or `0.48`. Observe that between evaluation iterations we also sample new goal positions / tasks.

Essentially you can think of it like "we sample `num_eval` number of 'meta batches' of tasks, we adapt the agent on each task, we then get the success rate of the agent for each task over `num_eval_episodes` episodes, then we average the success rate over all the tasks in a meta batch, and finally we average the success rate over all the meta batches". 
<details>
<summary>Explanation on meta-batch</summary>

What usually confuses people is how this meta batch is constructed in meta world. In ML1 it's simple, we simply sample goal positions for the given env. In ML10/45 it's a bit more confusing cause we want to somewhat evenly sample from the combinatorial `(env_type,goal_position)` space, which is further explained [here](https://metaworld.farama.org/evaluation/evaluation/#the-num-evals-parameter-and-meta-batch-size-in-metaworld).
</details>

Given this information we're hopefully now on the same page as to how evaluation *should* work in Metaworld, and I'm pretty certain it actually does work this way using the current code because it's what we used in the Meta-World+ paper, where indeed RL2 gets <50% success rate. Are you saying that currently running the RL2 example of this repository reports an overestimated number that seems wrong? I don't see how an env contributes "multiple times" to the success rate as you describe, because its *average* success rate is already returned by the `multitask_evaluation` function. Note that we also [sample new goal positions](https://github.com/Farama-Foundation/Metaworld/blob/master/metaworld/evaluation.py#L125) (new meta batch) each evaluation iteration, so it's not like it's the exact same task / env that's used next iteration, so it doesn't contribute multiple times that way either.

I can see an argument perhaps for ML10/45 for certain env types being easier to solve than others so even with new goal positions every iteration they still "contribute multiple times" in some sense but this is how metarl evaluation in Metaworld always worked (which you can verify by reviewing the original garage repo). However, the evaluation function already returns average success rate [per env type](https://github.com/Farama-Foundation/Metaworld/blob/master/metaworld/evaluation.py#L161) which I believe should be logged to wandb by default in this repo's training script, so if you want you can already inspect and monitor which env types do better / worse.

**Comment by yjy1231** [comment]
Thanks for the reply. I can see that the calculation in the function `metalearning_evaluation` is indeed `total_mean_success_rate / num_evals`, but this is not what I was asking. What I was asking was the way `success_rate_per_task` is calculated in each single `evaluation`. 

```
while not eval_done(episodic_returns):
        actions = agent.eval_action(obs)
        obs, _, terminations, truncations, infos = eval_envs.step(actions)

        dones = np.logical_or(terminations, truncations)
        agent.reset(dones)

        for i, env_ended in enumerate(dones):
            if env_ended:
                ...
                if len(episodic_returns[task_names[i]]) <= num_episodes:
                    successes[task_names[i]] += int(infos["final_info"]["success"][i])

    ...
    success_rate_per_task = {
        task_name: task_successes / num_episodes
        for task_name, task_successes in successes.items()
    }
    mean_success_rate = np.mean(list(success_rate_per_task.values()))
```

We can see that `successes` gets added by 1, if one environment has any success, no matter if that success has already been recorded or not. Once there's a success in the environment, that environment gets reset, and can possibly contribute to the success rate again. In ML1, as all testing environments are considered the same task, what happened is actually any successes in any testing environment would be counted in `successes` even repeatedly. Considering my previous example: a meta testing batch size of 10 corresponds to 10 different goal positions under the same task, and 1 of them was successful while 9 of them fail. The successful environment will keep contributing to `successes`, as it keeps getting reset after reaching the goal. Suppose `num_episodes` is 3, and within a 500 `max_step` the successful environment gets reset for 3 times. This is a 100% success rate following `task_successes / num_episodes`, but it is wrong, as the actual success rate is 10%. This is what causes huge overestimation of the success rate, no matter what method we are using.

**Comment by rainx0r** [comment]
Hmm I see, the issue is that "task_name" is the same in ML1 for all envs in the meta batch and it's actually used to count successes in the multitask evaluation. The actual env reset behaviour after success Is intended, because as I said we evaluate each task for N episodes, and also the `if len(episodic_returns[task_names[i]]) <= num_episodes:` guard is meant to discard any successes. that might have occurred after N episodes elapsed for some task. So the logic is correct and the bug is simply that we didn't consider the case where all the envs have the same task name because we mostly did ML10 and 45, where we also didn't observe any overestimation of success rate.

I'll make a PR to the metaworld repo that fixes this, thanks
